Add GBK charset description (#8101)
pingcap/docs-cn: character-set-gbk.md

---
title: GBK 字符集
---

# GBK 字符集

本文档介绍 TiDB 对 GBK 字符集的支持情况。

目前 TiDB 支持 GBK 字符集和排序规则情况：
```sql

> SHOW CHARACTER SET WHERE CHARSET = 'gbk';;

Reviewer: ```suggestion
SHOW CHARACTER SET WHERE CHARSET = 'gbk';
```

+---------+-------------------------------------+-------------------+--------+
| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_bin           |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+

> 2 rows in set (0.00 sec)

Reviewer: ```suggestion
1 rows in set (0.00 sec)
```

```
TiDB GBK 字符集默认排序规则与 MySQL 不一致，MySQL 的是 `gbk_chinese_ci`. 另外，目前 TiDB 支持的 `gbk_bin` 与 MySQL 使用的也不一致。TiDB 这边排序规则是在将 GBK 转换成 UTF8MB4 然后做二进制排序。
如果需要兼容 MySQL 的排序规则，需要使用[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)。

利用以下的语句可以查看字符集对应的排序规则（以下是[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)）下的结果：

```sql

> SHOW CHARACTER SET WHERE CHARSET = 'gbk';;

Reviewer 2: ```suggestion
SHOW CHARACTER SET WHERE CHARSET = 'gbk';
```

+---------+-------------------------------------+-------------------+--------+
| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_chinese_ci    |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
| gbk_chinese_ci | gbk     |   28 | Yes     | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
2 rows in set (0.00 sec)
```

## MySQL 兼容性

### 非法字符兼容性

> * 在设置成 `character_set_client = gbk`、`character_set_connection = utf8mb4` 和 `character_set_table = gbk` 的情况下，TiDB 处理非法字符的情况与 MySQL 兼容。

Reviewer 2: ```suggestion
* 在 `character_set_client` 和 `character_set_connection` 不都为 `gbk` 的情况下，TiDB 处理非法字符的情况与 MySQL 兼容。
```

> * 在设置成 `character_set_client = gbk`、`character_set_connection = utf8mb4` 和 `character_set_table = gbk` 的情况下，TiDB 处理非法字符的情况与 MySQL 兼容。

Reviewer 3: 这几个都是 MySQL 的配置项吗？

Author: 是，MySQL 的系统变量

> * 当设置 `names` 为 `gbk` 后 TiDB 处理非法字符的情况与 MySQL 有一些不一样。具体行为如下（其中 `insert` 语句所在表的建表语句：`create table gbk_table(a varchar(32) character set gbk);`）：

Reviewer 2: ```suggestion
* 在 `character_set_client` 和 `character_set_connection` 都为 `gbk` 的情况下， TiDB 处理非法字符的情况与 MySQL 有一些不一样。具体行为如下（其中 `insert` 语句所在表的建表语句：`create table gbk_table(a varchar(32) character set gbk);`）：
```

Author: 这边有 select 的结果，所以考虑的时候是 names 为 gbk
```
tidb> set names gbk;
Query OK, 0 rows affected (0.00 sec)
tidb> set character_set_client = utf8mb4;
Query OK, 0 rows affected (0.01 sec)
tidb> select hex('一a') ;
+------------+
| hex('һa')  |
+------------+
| D2BB61     |
+------------+
1 row in set (0.01 sec)

tidb> set character_set_client =gbk;
Query OK, 0 rows affected (0.00 sec)
tidb> select hex('一a') ;
ERROR 1300 (HY000): Invalid gbk character string: '8061'
```

Reviewer 2: `set names gbk `等价于 
```sql
set @@character_set_client=gbk
set @@character_set_connection=gbk
set @@character_set_results=gbk
```

所以实际上还是 
> 在 `character_set_client` 和 `character_set_connection` 都为 `gbk` 的情况下

Author: 嗯，我本来想说，这里和表格里面一样，所以这边写了 names。
已改。

---
title: GBK
---

# GBK

本文档介绍 TiDB 对 GBK 字符集的支持情况。

目前 TiDB 支持 GBK 字符集和排序规则情况：
```sql
SHOW CHARACTER SET WHERE CHARSET = 'gbk';
+---------+-------------------------------------+-------------------+--------+
| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_bin           |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
1 rows in set (0.00 sec)
```

> TiDB GBK 字符集默认排序规则与 MySQL 不一致，MySQL 的是 `gbk_chinese_ci`。 另外，目前 TiDB 支持的 `gbk_bin` 与 MySQL 使用的也不一致。TiDB 这边排序规则是在将 GBK 转换成 UTF8MB4 然后做二进制排序。
> 如果需要兼容 MySQL 的排序规则，需要使用[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)。

Reviewer 3: 两行之间需要有一个空行
```suggestion
TiDB GBK 字符集默认排序规则与 MySQL 不一致，MySQL 的是 `gbk_chinese_ci`。 另外，目前 TiDB 支持的 `gbk_bin` 与 MySQL 使用的也不一致。TiDB 这边排序规则是在将 GBK 转换成 UTF8MB4 然后做二进制排序。

如果需要兼容 MySQL 的排序规则，需要使用[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)。
```

利用以下的语句可以查看字符集对应的排序规则（以下是[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)）下的结果：

```sql
SHOW CHARACTER SET WHERE CHARSET = 'gbk';;
+---------+-------------------------------------+-------------------+--------+
| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_chinese_ci    |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
| gbk_chinese_ci | gbk     |   28 | Yes     | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
2 rows in set (0.00 sec)
```

> 

Reviewer 3: QA 的 Compatibility change list 里还写了这些兼容性限制，不知是否要加进这篇文档中：

    - TiFlash 暂不支持 GBK 字符。
    - TiCDC 暂不支持 GBK 字符。
    - 如果 `character_set_client` 和 `character_set_connection` 都是 `gbk` 时，处理非法 GBK 字符与 MySQL 存在兼容性问题。
    - `character_set_client` 在处理 `prepare` 语句时可能出现兼容性问题。
    - TiDB 不支持 `_gbk"xxx"` 的用法，但是支持 `_utf8mb4"xxx"` 的用法。而 MySQL 对于 `_charset"xxx"` 的用法都支持。
    - TiDB Lightning 在 v5.4.0 之前不支持导入 `charset=GBK` 的表。BR 在 v5.3.0 之前不支持恢复 `charset=GBK` 的表。

| TiDB    | not include          | select hex('一a') (0xe4b88061)       | select hex('一a') (0xe4b88061)        |
|         | STRICT_ALL_TABLES or | e6b6933f                             | e4b83f                               |
|         | STRICT_TRANS_TABLES  |                                      |                                      |
|         |                      | insert into gbk_table values('一a'); | insert into gbk_table values('一a');  |
|         |                      | select hex(a) from gbk_table;        | select hex(a) from gbk_table;        |
|         |                      | e4b83f                               | e4b83f                               |
+---------+----------------------+--------------------------------------+--------------------------------------+
```

### 其它

* 目前不支持通过 `alter table` 语句将其它字符集类型改成 `gbk`，或者从 `gbk` 转成其它字符集。

* 不支持使用 `_gbk`， 比如：

  ```sql
  create table t(a char(10) charset binary);
  Query OK, 0 rows affected (0.00 sec)
  insert into t values (_gbk'啊');
  ERROR 1115 (42000): Unsupported character introducer: 'gbk'
  ```

* 对于 `ENUM` 类型中的二进制字符，目前都会将其作为 utf8mb4 字符集处理。

> ## 其他组件兼容性

Reviewer 3: QA 的 Compatibility change list 里还写了这些兼容性限制，不知是否要加进这篇文档中：

    - TiFlash 暂不支持 GBK 字符。
    - TiCDC 暂不支持 GBK 字符。
    - 如果 `character_set_client` 和 `character_set_connection` 都是 `gbk` 时，处理非法 GBK 字符与 MySQL 存在兼容性问题。
    - `character_set_client` 在处理 `prepare` 语句时可能出现兼容性问题。
    - TiDB 不支持 `_gbk"xxx"` 的用法，但是支持 `_utf8mb4"xxx"` 的用法。而 MySQL 对于 `_charset"xxx"` 的用法都支持。
    - TiDB Lightning 在 v5.4.0 之前不支持导入 `charset=GBK` 的表。BR 在 v5.3.0 之前不支持恢复 `charset=GBK` 的表。

Author: 1. 这个是 mysql 本身的问题，所以没有在 user docs 带上
2. 已经带了
3. lighting 和 DM 都是需要和 TiDB 相同版本才能支持，这个不需特别说明吧?

Reviewer 4: 3. 我觉得需要特别说明，因为用户的 lightning 和 DM 版本可能和 TiDB 版本不同。@TomShawn 觉得呢

Reviewer 4: 已经加了。谢谢 @zimulala

---

> title: GBK

Author: Done.

---

# GBK

本文档介绍 TiDB 对 GBK 字符集的支持情况。

目前 TiDB 支持 GBK 字符集和排序规则情况：
```sql
SHOW CHARACTER SET WHERE CHARSET = 'gbk';
+---------+-------------------------------------+-------------------+--------+
| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_bin           |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
1 rows in set (0.00 sec)
```

> TiDB GBK 字符集默认排序规则与 MySQL 不一致，MySQL 的是 `gbk_chinese_ci`。 另外，目前 TiDB 支持的 `gbk_bin` 与 MySQL 使用的也不一致。TiDB 这边排序规则是在将 GBK 转换成 UTF8MB4 然后做二进制排序。

Reviewer 4: ```suggestion
TiDB GBK 字符集默认排序规则与 MySQL 不一致，TiDB GBK 字符集的默认排序规则为 `gbk_bin`，MySQL 的字符集默认排序规则是 `gbk_chinese_ci`。 另外，目前 TiDB 支持的 `gbk_bin` 与 MySQL 支持的 `gbk_bin` 也不一致，TiDB 是将 GBK 转换成 UTF8MB4 然后做二进制排序。
```

> 如果需要兼容 MySQL 的排序规则，需要使用[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)。

Reviewer 4: ```suggestion
如果要兼容 MySQL 的 GBK 字符集排序规则，你需要在初次初始化集群时通过设置 TiDB 配置项[`new_collations_enabled_on_first_bootstrap`](/tidb-configuration-file.md#new_collations_enabled_on_first_bootstrap) 为 `true`来开启[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)。
```

> 利用以下的语句可以查看字符集对应的排序规则（以下是[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)）下的结果：

Reviewer 4: ```suggestion
开启新的排序规则框架后，你可以通过以下 SQL 语句查看 GBK 字符集对应的排序规则：
```

```sql

> SHOW CHARACTER SET WHERE CHARSET = 'gbk';;

Reviewer 4: ```suggestion
SHOW CHARACTER SET WHERE CHARSET = 'gbk';
```

+---------+-------------------------------------+-------------------+--------+
| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_chinese_ci    |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
| gbk_chinese_ci | gbk     |   28 | Yes     | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
2 rows in set (0.00 sec)
```

## MySQL 兼容性

### 非法字符兼容性

> * 在 `character_set_client` 和 `character_set_connection` 不都为 `gbk` 的情况下，TiDB 处理非法字符的情况与 MySQL 兼容。

Reviewer 4: ```suggestion
* 在 `character_set_client` 和 `character_set_connection` 不同时为 `gbk` 的情况下，TiDB 处理非法字符的方式与 MySQL 一致。
```

> * 在 `character_set_client` 和 `character_set_connection` 都为 `gbk` 的情况下， TiDB 处理非法字符的情况与 MySQL 有一些不一样。例如，以 `set names gbk` 为前提，具体行为如下（其中 `insert` 语句所在表的建表语句：`create table gbk_table(a varchar(32) character set gbk);`）：

Reviewer 4: ```suggestion
* 在 `character_set_client` 和 `character_set_connection` 同时为 `gbk` 的情况下， TiDB 处理非法字符的方式与 MySQL 有所区别。例如，以 `set names gbk` 为前提，具体行为如下（其中 `insert` 语句所在表的建表语句：`create table gbk_table(a varchar(32) character set gbk);`）：
```

| DB    |    sql_mode:<br><br>  STRICT_ALL_TABLES or<br>  STRICT_TRANS_TABLES                                               | sql_mode 不包括： <br><br>  STRICT_ALL_TABLES or<br>  STRICT_TRANS_TABLES                                                                     |
|-------|-------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------|
| MySQL | `select hex('一a');` (0xe4b88061)<br> e4b88061<br><br>`insert into gbk_table values('一a');`<br> Incorrect Error       | `select hex('一a');` (0xe4b88061)<br> e4b88061<br><br>`insert into gbk_table values('一a');`<br>`select hex(a) from gbk_table;`<br> e4b8 |
| TiDB  | `select hex('一a');` (0xe4b88061)<br> Incorrect Error<br><br>`insert into gbk_table values('一a');`<br> Incorrect Error | `select hex('一a');` (0xe4b88061)<br> e4b83f<br><br>`insert into gbk_table values('一a');`<br>`select hex(a) from gbk_table;`<br> e4b83f  |

### 其它

* 目前不支持通过 `alter table` 语句将其它字符集类型改成 `gbk`，或者从 `gbk` 转成其它字符集。

* 不支持使用 `_gbk`， 比如：

  ```sql

>   create table t(a char(10) charset binary);

Reviewer 4: 请将本 PR 中所有新增 SQL 语句中的小写关键字改为大写，原因可参考https://github.com/pingcap/docs/issues/699

---
title: GBK
summary: 本文介绍 TiDB 对 GBK 字符集的支持情况。
---

# GBK

> 本文档介绍 TiDB 对 GBK 字符集的支持和兼容性情况。
> 
> 从 TiDB v5.4.0 起，TiDB 支持 GBK 字符集。

Reviewer 4: ```suggestion
TiDB 从 v5.4.0 开始支持 GBK 字符集。本文档介绍 TiDB 对 GBK 字符集的支持和兼容性情况。
```

```sql
SHOW CHARACTER SET WHERE CHARSET = 'gbk';
+---------+-------------------------------------+-------------------+--------+
| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_bin           |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
1 rows in set (0.00 sec)
```

> TiDB GBK 字符集默认排序规则与 MySQL 不一致，TiDB GBK 字符集的默认排序规则为 `gbk_bin`，MySQL 的字符集默认排序规则是 `gbk_chinese_ci`。 另外，目前 TiDB 支持的 `gbk_bin` 与 MySQL 支持的 `gbk_bin` 也不一致，TiDB 是将 GBK 转换成 UTF8MB4 然后做二进制排序。

Reviewer 4: 可以问下为啥 TiDB 要使用 `gbk_bin` 而不是 `gbk_chinese_ci` 作为 GBK 的默认排序规则吗？ `gbk_bin` 有啥优势吗？

> TiDB GBK 字符集默认排序规则与 MySQL 不一致，TiDB GBK 字符集的默认排序规则为 `gbk_bin`，MySQL 的字符集默认排序规则是 `gbk_chinese_ci`。 另外，目前 TiDB 支持的 `gbk_bin` 与 MySQL 支持的 `gbk_bin` 也不一致，TiDB 是将 GBK 转换成 UTF8MB4 然后做二进制排序。

Author: 不是有优势，只是没有开启“新的排序规则框架”时，TiDB 原来的框架里面支持的也都是“_bin”排序，一是兼容，二是旧框架实现起来也很复杂且必要性不强

> 如果要兼容 MySQL 的 GBK 字符集排序规则，你需要在初次初始化集群时通过设置 TiDB 配置项[`new_collations_enabled_on_first_bootstrap`](/tidb-configuration-file.md#new_collations_enabled_on_first_bootstrap) 为 `true`来开启[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)。

Reviewer 4: ```suggestion
如果要使 TiDB 兼容 MySQL 的 GBK 字符集排序规则，你需要在初次初始化 TiDB 集群时设置 TiDB 配置项[`new_collations_enabled_on_first_bootstrap`](/tidb-configuration-file.md#new_collations_enabled_on_first_bootstrap) 为 `true`来开启[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)。
```

> 开启新的排序规则框架后，你可以通过以下 SQL 语句查看 GBK 字符集对应的排序规则：

Reviewer 4: ```suggestion
开启新的排序规则框架后，如果再次查看 GBK 字符集对应的排序规则，你可以看到 TiDB GBK 默认排序规则已经切换为 `gbk_chinese_ci `。
```

```sql
SHOW CHARACTER SET WHERE CHARSET = 'gbk';
+---------+-------------------------------------+-------------------+--------+
| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_chinese_ci    |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
| gbk_chinese_ci | gbk     |   28 | Yes     | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
2 rows in set (0.00 sec)
```

## MySQL 兼容性

### 非法字符兼容性

> * 在 `character_set_client` 和 `character_set_connection` 不同时为 `gbk` 的情况下，TiDB 处理非法字符的方式与 MySQL 一致。

Reviewer 4: ```suggestion
* 在系统变量 [`character_set_client`](/system-variables/#character_set_client) 和 [`character_set_connection`](/system-variables/#character_set_connection) 不同时设置为 `gbk` 的情况下，TiDB 处理非法字符的方式与 MySQL 一致。
```

> * 在 `character_set_client` 和 `character_set_connection` 同时为 `gbk` 的情况下， TiDB 处理非法字符的方式与 MySQL 有所区别。例如，以 `set names gbk` 为前提，具体行为如下（其中 `insert` 语句所在表的建表语句：`create table gbk_table(a varchar(32) character set gbk);`）：

Reviewer 4: ```suggestion
* 在 `character_set_client` 和 `character_set_connection` 同时为 `gbk` 的情况下， TiDB 处理非法字符的方式与 MySQL 有所区别。

    例如，当 `SET NAMES gbk` 时，如果分别在 MySQL 和 TiDB 上通过 `CREATE TABLE gbk_table(A VARCHAR(32) CHARACTER SET gbk);` 语句建表，然后按照下表中的 SQL 语句进行操作，就能看到具体的区别。
```

> | DB    |    sql_mode:<br><br>  STRICT_ALL_TABLES or<br>  STRICT_TRANS_TABLES                                               | sql_mode 不包括： <br><br>  STRICT_ALL_TABLES or<br>  STRICT_TRANS_TABLES                                                                     |

Reviewer 4: ```suggestion
| DB    |    如何设置的 SQL 模式包含 `STRICT_ALL_TABLES` 或 `STRICT_TRANS_TABLES`                                               | 如果设置的 SQL 模式不包含 `STRICT_ALL_TABLES` 和  `STRICT_TRANS_TABLES`                                                                     |
```

|-------|-------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------|
| MySQL | `select hex('一a');` (0xe4b88061)<br> e4b88061<br><br>`insert into gbk_table values('一a');`<br> Incorrect Error       | `select hex('一a');` (0xe4b88061)<br> e4b88061<br><br>`insert into gbk_table values('一a');`<br>`select hex(a) from gbk_table;`<br> e4b8 |
| TiDB  | `select hex('一a');` (0xe4b88061)<br> Incorrect Error<br><br>`insert into gbk_table values('一a');`<br> Incorrect Error | `select hex('一a');` (0xe4b88061)<br> e4b83f<br><br>`insert into gbk_table values('一a');`<br>`select hex(a) from gbk_table;`<br> e4b83f  |

### 其它

* 目前 TiDB 不支持通过 `ALTER TABLE` 语句将其它字符集类型改成 `gbk`，或者从 `gbk` 转成其它字符集类型。

* TiDB 不支持使用 `_gbk`， 比如：

  ```sql
  CREATE TABLE t(a CHAR(10) CHARSET BINARY);
  Query OK, 0 rows affected (0.00 sec)
  INSERT INTO t VALUES (_gbk'啊');
  ERROR 1115 (42000): Unsupported character introducer: 'gbk'
  ```

> * 对于 `ENUM` 类型中的二进制字符，TiDB 目前都会将其作为 `utf8mb4` 字符集处理。

Reviewer 2: `enum` 和 `set` 都是这样的吧 /cc @tangenta

Reviewer 4: @zimulala 麻烦请加下吧，谢谢

Author: Done.

> * 对于 `ENUM` 类型中的二进制字符，TiDB 目前都会将其作为 `utf8mb4` 字符集处理。

Author: ```suggestion
* 对于 `ENUM` 和 `SET` 类型中的二进制字符，TiDB 目前都会将其作为 `utf8mb4` 字符集处理。
```

---
title: GBK
summary: 本文介绍 TiDB 对 GBK 字符集的支持情况。
---

# GBK

> 本文档介绍 TiDB 对 GBK 字符集的支持和兼容性情况。
> 

Reviewer 4: ```suggestion
```

| gbk_chinese_ci | gbk     |   28 | Yes     | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
2 rows in set (0.00 sec)
```

## MySQL 兼容性

### 非法字符兼容性

* 在系统变量 [`character_set_client`](/system-variables/#character_set_client) 和 [`character_set_connection`](/system-variables/#character_set_connection) 不同时设置为 `gbk` 的情况下，TiDB 处理非法字符的方式与 MySQL 一致。
* 在 `character_set_client` 和 `character_set_connection` 同时为 `gbk` 的情况下， TiDB 处理非法字符的方式与 MySQL 有所区别。

      例如，当 `SET NAMES gbk` 时，如果分别在 MySQL 和 TiDB 上通过 `CREATE TABLE gbk_table(a VARCHAR(32) CHARACTER SET gbk)` 语句建表，然后按照下表中的 SQL 语句进行操作，就能看到具体的区别。

| DB    |    如果设置的 SQL 模式包含 `STRICT_ALL_TABLES` 或 `STRICT_TRANS_TABLES`                                               | 如果设置的 SQL 模式不包含 `STRICT_ALL_TABLES` 和  `STRICT_TRANS_TABLES`                                                                     |
|-------|-------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------|

> | MySQL | `select hex('一a');` (0xe4b88061)<br> e4b88061<br><br>`insert into gbk_table values('一a');`<br> Incorrect Error       | `select hex('一a');` (0xe4b88061)<br> e4b88061<br><br>`insert into gbk_table values('一a');`<br>`select hex(a) from gbk_table;`<br> e4b8 |

Reviewer 4: ```suggestion
| MySQL | `SELECT HEX('一a');` (0xe4b88061)<br> e4b88061<br><br>`INSERT INTO gbk_table values('一a');`<br> Incorrect Error       | `SELECT HEX('一a');` (0xe4b88061)<br> e4b88061<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> e4b8 |
```

> | MySQL | `SELECT HEX('一a');` (0xe4b88061)<br> e4b88061<br><br>`INSERT INTO gbk_table values('一a');`<br> Incorrect Error       | `SELECT HEX('一a');` (0xe4b88061)<br> e4b88061<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> e4b8 |
> | TiDB  | `SELECT HEX('一a');` (0xe4b88061)<br> Incorrect Error<br><br>`INSERT INTO gbk_table VALUES('一a');`<br> Incorrect Error | `SELECT HEX('一a');` (0xe4b88061)<br> e4b83f<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> e4b83f  |

Reviewer 4: ```suggestion
| MySQL | `SELECT HEX('一a');` (0xe4b88061)<br> `e4b88061`<br><br>`INSERT INTO gbk_table values('一a');`<br> `Incorrect Error`       | `SELECT HEX('一a');` (0xe4b88061)<br> `e4b88061`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> `e4b8` |
| TiDB  | `SELECT HEX('一a');` (0xe4b88061)<br> `Incorrect Error`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br> `Incorrect Error` | `SELECT HEX('一a');` (0xe4b88061)<br> `e4b83f`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> `e4b83f`  |
```

> | TiDB  | `select hex('一a');` (0xe4b88061)<br> Incorrect Error<br><br>`insert into gbk_table values('一a');`<br> Incorrect Error | `select hex('一a');` (0xe4b88061)<br> e4b83f<br><br>`insert into gbk_table values('一a');`<br>`select hex(a) from gbk_table;`<br> e4b83f  |

Reviewer 4: ```suggestion
| TiDB  | `SELECT HEX('一a');` (0xe4b88061)<br> Incorrect Error<br><br>`INSERT INTO gbk_table VALUES('一a');`<br> Incorrect Error | `SELECT HEX('一a');` (0xe4b88061)<br> e4b83f<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> e4b83f  |
```

> | MySQL | `SELECT HEX('一a');` (0xe4b88061)<br> `e4b88061`<br><br>`INSERT INTO gbk_table values('一a');`<br> `Incorrect Error`       | `SELECT HEX('一a');` (0xe4b88061)<br> `e4b88061`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> `e4b8` |
> | TiDB  | `SELECT HEX('一a');` (0xe4b88061)<br> `Incorrect Error`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br> `Incorrect Error` | `SELECT HEX('一a');` (0xe4b88061)<br> `e4b83f`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> `e4b83f`  |

Reviewer 4: ```suggestion
| MySQL | `SELECT HEX('一a');` <br> `e4b88061`<br><br>`INSERT INTO gbk_table values('一a');`<br> `Incorrect Error`       | `SELECT HEX('一a');` <br> `e4b88061`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> `e4b8` |
| TiDB  | `SELECT HEX('一a');` <br> `Incorrect Error`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br> `Incorrect Error` | `SELECT HEX('一a');` <br> `e4b83f`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> `e4b83f`  |
```

### 其它

* 目前 TiDB 不支持通过 `ALTER TABLE` 语句将其它字符集类型改成 `gbk`，或者从 `gbk` 转成其它字符集类型。

* TiDB 不支持使用 `_gbk`， 比如：

  ```sql
  CREATE TABLE t(a CHAR(10) CHARSET BINARY);
  Query OK, 0 rows affected (0.00 sec)
  INSERT INTO t VALUES (_gbk'啊');
  ERROR 1115 (42000): Unsupported character introducer: 'gbk'
  ```

* 对于 `ENUM` 和 `SET` 类型中的二进制字符，TiDB 目前都会将其作为 `utf8mb4` 字符集处理。

## 其他组件兼容性

* 目前不支持 GBK 字符集的 TiDB 组件包括：TiCDC 和 TiFlash。

> * TiDB Lightning 在 v5.4.0 之前不支持导入 `charset=GBK` 的表。BR 在 v5.4.0 之前不支持恢复 `charset=GBK` 的表。

Author: ```suggestion
* TiDB Lightning 在 v5.4.0 之前不支持导入 `charset=GBK` 的表。BR 在 v5.3.0 之前不支持恢复 `charset=GBK` 的表。
```

例如，当 `SET NAMES gbk` 时，如果分别在 MySQL 和 TiDB 上通过 `CREATE TABLE gbk_table(a VARCHAR(32) CHARACTER SET gbk)` 语句建表，然后按照下表中的 SQL 语句进行操作，就能看到具体的区别。

| 数据库    |    如果设置的 SQL 模式包含 `STRICT_ALL_TABLES` 或 `STRICT_TRANS_TABLES`                                               | 如果设置的 SQL 模式不包含 `STRICT_ALL_TABLES` 和  `STRICT_TRANS_TABLES`                                                                     |
|-------|-------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------|
| MySQL | `SELECT HEX('一a');` <br> `e4b88061`<br><br>`INSERT INTO gbk_table values('一a');`<br> `Incorrect Error`       | `SELECT HEX('一a');` <br> `e4b88061`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> `e4b8` |
| TiDB  | `SELECT HEX('一a');` <br> `Incorrect Error`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br> `Incorrect Error` | `SELECT HEX('一a');` <br> `e4b83f`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> `e4b83f`  |

> 说明：该表中 `HEX('一a')` 的转换为十六进制的真实结果为 `0xe4b88061`。

Reviewer 4: ```suggestion
说明：该表中 `HEX('一a')` 转换为十六进制的真实结果为 `0xe4b88061`。

```

### 其它

* 目前 TiDB 不支持通过 `ALTER TABLE` 语句将其它字符集类型改成 `gbk`，或者从 `gbk` 转成其它字符集类型。

* TiDB 不支持使用 `_gbk`， 比如：

  ```sql
  CREATE TABLE t(a CHAR(10) CHARSET BINARY);
  Query OK, 0 rows affected (0.00 sec)
  INSERT INTO t VALUES (_gbk'啊');
  ERROR 1115 (42000): Unsupported character introducer: 'gbk'
  ```

* 对于 `ENUM` 和 `SET` 类型中的二进制字符，TiDB 目前都会将其作为 `utf8mb4` 字符集处理。

> ## 其他组件兼容性

Reviewer 4: ```suggestion

## 组件兼容性
```

| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_bin           |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)
SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
1 rows in set (0.00 sec)
```

## 排序规则

MySQL 的字符集默认排序规则是 `gbk_chinese_ci`。与 MySQL 不同，TiDB GBK 字符集的默认排序规则为 `gbk_bin`。 另外，TiDB 支持的 `gbk_bin` 与 MySQL 支持的 `gbk_bin` 排序规则也不一致，TiDB 是将 GBK 转换成 UTF8MB4 然后做二进制排序。

如果要使 TiDB 兼容 MySQL 的 GBK 字符集排序规则，你需要在初次初始化 TiDB 集群时设置 TiDB 配置项[`new_collations_enabled_on_first_bootstrap`](/tidb-configuration-file.md#new_collations_enabled_on_first_bootstrap) 为 `true` 来开启[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)。

> 开启新的排序规则框架后，如果查看 GBK 字符集对应的排序规则，你可以看到 TiDB GBK 默认排序规则已经切换为 `gbk_chinese_ci `。

Reviewer 5: ```suggestion
开启新的排序规则框架后，如果查看 GBK 字符集对应的排序规则，你可以看到 TiDB GBK 默认排序规则已经切换为 `gbk_chinese_ci`。
```

---
title: GBK
summary: 本文介绍 TiDB 对 GBK 字符集的支持情况。
---

# GBK

TiDB 从 v5.4.0 开始支持 GBK 字符集。本文档介绍 TiDB 对 GBK 字符集的支持和兼容情况。

```sql
SHOW CHARACTER SET WHERE CHARSET = 'gbk';
+---------+-------------------------------------+-------------------+--------+
| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_bin           |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

> SHOW COLLATION WHERE CHARSET = 'gbk';

Reviewer: ```suggestion

SHOW COLLATION WHERE CHARSET = 'gbk';
```

+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
1 rows in set (0.00 sec)
```

## 排序规则

> MySQL 的字符集默认排序规则是 `gbk_chinese_ci`。与 MySQL 不同，TiDB GBK 字符集的默认排序规则为 `gbk_bin`。 另外，TiDB 支持的 `gbk_bin` 与 MySQL 支持的 `gbk_bin` 排序规则也不一致，TiDB 是将 GBK 转换成 UTF8MB4 然后做二进制排序。

Reviewer: ```suggestion
MySQL 的字符集默认排序规则是 `gbk_chinese_ci`。与 MySQL 不同，TiDB GBK 字符集的默认排序规则为 `gbk_bin`。另外，TiDB 支持的 `gbk_bin` 与 MySQL 支持的 `gbk_bin` 排序规则也不一致，TiDB 是将 GBK 转换成 UTF8MB4 然后做二进制排序。
```

```sql
SHOW CHARACTER SET WHERE CHARSET = 'gbk';
+---------+-------------------------------------+-------------------+--------+
| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_chinese_ci    |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
| gbk_chinese_ci | gbk     |   28 | Yes     | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
2 rows in set (0.00 sec)
```

## MySQL 兼容性

### 非法字符兼容性

* 在系统变量 [`character_set_client`](/system-variables.md#character_set_client) 和 [`character_set_connection`](/system-variables.md#character_set_connection) 不同时设置为 `gbk` 的情况下，TiDB 处理非法字符的方式与 MySQL 一致。

> * 在 `character_set_client` 和 `character_set_connection` 同时为 `gbk` 的情况下， TiDB 处理非法字符的方式与 MySQL 有所区别。MySQL 处理非法 GBK 字符集时，对读和写操作的处理方式不一样，但在 TiDB 中处理方式是一样的。即 TiDB 在严格模式下读写非法 GBK 字符都会报错，在非严格模式下，读写非法 GBK 字符都会用 `?` 替换。

Reviewer 4: ```suggestion
* 在 `character_set_client` 和 `character_set_connection` 同时为 `gbk` 的情况下， TiDB 处理非法字符的方式与 MySQL 有所区别。

    - MySQL 处理非法 GBK 字符集时，对读和写操作的处理方式不同。
    - TiDB 处理非法 GBK 字符集时，对读和写操作的处理方式相同。TiDB 在严格模式下读写非法 GBK 字符都会报错，在非严格模式下，读写非法 GBK 字符都会用 `?` 替换。
```

| MySQL | `SELECT HEX('一a');` <br /> `e4b88061`<br /><br />`INSERT INTO gbk_table values('一a');`<br /> `Incorrect Error`       | `SELECT HEX('一a');` <br /> `e4b88061`<br /><br />`INSERT INTO gbk_table VALUES('一a');`<br />`SELECT HEX(a) FROM gbk_table;`<br /> `e4b8` |
| TiDB  | `SELECT HEX('一a');` <br /> `Incorrect Error`<br /><br />`INSERT INTO gbk_table VALUES('一a');`<br /> `Incorrect Error` | `SELECT HEX('一a');` <br /> `e4b83f`<br /><br />`INSERT INTO gbk_table VALUES('一a');`<br />`SELECT HEX(a) FROM gbk_table;`<br /> `e4b83f`  |

说明：该表中 `SELECT HEX('一a');` 在 `utf8mb4` 字节集下的结果为 `e4b88061`。

### 其它

* 目前 TiDB 不支持通过 `ALTER TABLE` 语句将其它字符集类型改成 `gbk` 或者从 `gbk` 转成其它字符集类型。

* TiDB 不支持使用 `_gbk`， 比如：

  ```sql
  CREATE TABLE t(a CHAR(10) CHARSET BINARY);
  Query OK, 0 rows affected (0.00 sec)
  INSERT INTO t VALUES (_gbk'啊');
  ERROR 1115 (42000): Unsupported character introducer: 'gbk'
  ```

* 对于 `ENUM` 和 `SET` 类型中的二进制字符，TiDB 目前都会将其作为 `utf8mb4` 字符集处理。

> 
> 

Reviewer 5: ```suggestion

```

---
title: GBK
summary: 本文介绍 TiDB 对 GBK 字符集的支持情况。
---

# GBK

TiDB 从 v5.4.0 开始支持 GBK 字符集。本文档介绍 TiDB 对 GBK 字符集的支持和兼容性情况。

```sql
SHOW CHARACTER SET WHERE CHARSET = 'gbk';
+---------+-------------------------------------+-------------------+--------+
| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_bin           |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
1 rows in set (0.00 sec)
```

## 排序规则

MySQL 的字符集默认排序规则是 `gbk_chinese_ci`。与 MySQL 不同，TiDB GBK 字符集的默认排序规则为 `gbk_bin`。 另外，TiDB 支持的 `gbk_bin` 与 MySQL 支持的 `gbk_bin` 排序规则也不一致，TiDB 是将 GBK 转换成 UTF8MB4 然后做二进制排序。

> 如果要使 TiDB 兼容 MySQL 的 GBK 字符集排序规则，你需要在初次初始化 TiDB 集群时设置 TiDB 配置项[`new_collations_enabled_on_first_bootstrap`](/tidb-configuration-file.md#new_collations_enabled_on_first_bootstrap) 为 `true`来开启[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)。

Reviewer 4: ```suggestion
如果要使 TiDB 兼容 MySQL 的 GBK 字符集排序规则，你需要在初次初始化 TiDB 集群时设置 TiDB 配置项[`new_collations_enabled_on_first_bootstrap`](/tidb-configuration-file.md#new_collations_enabled_on_first_bootstrap) 为 `true` 来开启[新的排序规则框架](/character-set-and-collation.md#新框架下的排序规则支持)。
```

开启新的排序规则框架后，如果查看 GBK 字符集对应的排序规则，你可以看到 TiDB GBK 默认排序规则已经切换为 `gbk_chinese_ci `。

```sql
SHOW CHARACTER SET WHERE CHARSET = 'gbk';
+---------+-------------------------------------+-------------------+--------+
| Charset | Description                         | Default collation | Maxlen |
+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_chinese_ci    |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
| gbk_chinese_ci | gbk     |   28 | Yes     | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
2 rows in set (0.00 sec)
```

## MySQL 兼容性

### 非法字符兼容性

> * 在系统变量 [`character_set_client`](/system-variables/#character_set_client) 和 [`character_set_connection`](/system-variables/#character_set_connection) 不同时设置为 `gbk` 的情况下，TiDB 处理非法字符的方式与 MySQL 一致。

Reviewer 3: ```suggestion
* 在系统变量 [`character_set_client`](/system-variables.md#character_set_client) 和 [`character_set_connection`](/system-variables.md#character_set_connection) 不同时设置为 `gbk` 的情况下，TiDB 处理非法字符的方式与 MySQL 一致。
```

> * 在 `character_set_client` 和 `character_set_connection` 同时为 `gbk` 的情况下， TiDB 处理非法字符的方式与 MySQL 有所区别。

Reviewer 4: @zimulala 可以看看能够描述下整体的区别吗？

> * 在 `character_set_client` 和 `character_set_connection` 同时为 `gbk` 的情况下， TiDB 处理非法字符的方式与 MySQL 有所区别。

Author: Done.

> * 在 `character_set_client` 和 `character_set_connection` 同时为 `gbk` 的情况下， TiDB 处理非法字符的方式与 MySQL 有所区别。
> 
> 例如，当 `SET NAMES gbk` 时，如果分别在 MySQL 和 TiDB 上通过 `CREATE TABLE gbk_table(a VARCHAR(32) CHARACTER SET gbk)` 语句建表，然后按照下表中的 SQL 语句进行操作，就能看到具体的区别。

Reviewer 4: ```suggestion
* 在 `character_set_client` 和 `character_set_connection` 同时为 `gbk` 的情况下， TiDB 处理非法字符的方式与 MySQL 有所区别。例如，当 `SET NAMES gbk` 时，如果分别在 MySQL 和 TiDB 上通过 `CREATE TABLE gbk_table(a VARCHAR(32) CHARACTER SET gbk)` 语句建表，然后按照下表中的 SQL 语句进行操作，就能看到输出的区别。
```

> | 数据库    |    如果设置的 SQL 模式包含 `STRICT_ALL_TABLES` 或 `STRICT_TRANS_TABLES`                                               | 如果设置的 SQL 模式不包含 `STRICT_ALL_TABLES` 和  `STRICT_TRANS_TABLES`                                                                     |

Reviewer 5: ```suggestion
| 数据库    |    SQL 模式包含 `STRICT_ALL_TABLES` 或 `STRICT_TRANS_TABLES`                                               |  SQL 模式不包含 `STRICT_ALL_TABLES` 和  `STRICT_TRANS_TABLES`                                                                     |
```
“包含 A 或 B”，“不包含 A 和 B”，“和”“或”是专门这么写的是吗？

Reviewer 4: 是的。

|-------|-------------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------|

> | MySQL | `SELECT HEX('一a');` <br> `e4b88061`<br><br>`INSERT INTO gbk_table values('一a');`<br> `Incorrect Error`       | `SELECT HEX('一a');` <br> `e4b88061`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> `e4b8` |
> | TiDB  | `SELECT HEX('一a');` <br> `Incorrect Error`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br> `Incorrect Error` | `SELECT HEX('一a');` <br> `e4b83f`<br><br>`INSERT INTO gbk_table VALUES('一a');`<br>`SELECT HEX(a) FROM gbk_table;`<br> `e4b83f`  |

Reviewer 3: ```suggestion
| MySQL | `SELECT HEX('一a');` <br /> `e4b88061`<br /><br />`INSERT INTO gbk_table values('一a');`<br /> `Incorrect Error`       | `SELECT HEX('一a');` <br /> `e4b88061`<br /><br />`INSERT INTO gbk_table VALUES('一a');`<br />`SELECT HEX(a) FROM gbk_table;`<br /> `e4b8` |
| TiDB  | `SELECT HEX('一a');` <br /> `Incorrect Error`<br /><br />`INSERT INTO gbk_table VALUES('一a');`<br /> `Incorrect Error` | `SELECT HEX('一a');` <br /> `e4b83f`<br /><br />`INSERT INTO gbk_table VALUES('一a');`<br />`SELECT HEX(a) FROM gbk_table;`<br /> `e4b83f`  |
```

说明：该表中 `HEX('一a')` 转换为十六进制的真实结果为 `0xe4b88061`。

### 其它

> * 目前 TiDB 不支持通过 `ALTER TABLE` 语句将其它字符集类型改成 `gbk`，或者从 `gbk` 转成其它字符集类型。

Reviewer 5: ```suggestion
* 目前 TiDB 不支持通过 `ALTER TABLE` 语句将其它字符集类型改成 `gbk` 或者从 `gbk` 转成其它字符集类型。
```

* TiDB 不支持使用 `_gbk`， 比如：

  ```sql
  CREATE TABLE t(a CHAR(10) CHARSET BINARY);
  Query OK, 0 rows affected (0.00 sec)
  INSERT INTO t VALUES (_gbk'啊');
  ERROR 1115 (42000): Unsupported character introducer: 'gbk'
  ```

* 对于 `ENUM` 和 `SET` 类型中的二进制字符，TiDB 目前都会将其作为 `utf8mb4` 字符集处理。

## 组件兼容性

> * 目前不支持 GBK 字符集的 TiDB 组件包括：TiCDC 和 TiFlash。

Reviewer 5: ```suggestion
* TiCDC 和 TiFlash 目前不支持 GBK 字符集。
```
跟下面的风格保持一致，直接工具开头。

+---------+-------------------------------------+-------------------+--------+
| gbk     | Chinese Internal Code Specification | gbk_chinese_ci    |      2 |
+---------+-------------------------------------+-------------------+--------+
1 row in set (0.00 sec)

SHOW COLLATION WHERE CHARSET = 'gbk';
+----------------+---------+------+---------+----------+---------+
| Collation      | Charset | Id   | Default | Compiled | Sortlen |
+----------------+---------+------+---------+----------+---------+
| gbk_bin        | gbk     |   87 |         | Yes      |       1 |
| gbk_chinese_ci | gbk     |   28 | Yes     | Yes      |       1 |
+----------------+---------+------+---------+----------+---------+
2 rows in set (0.00 sec)
```

## MySQL 兼容性

### 非法字符兼容性

* 在系统变量 [`character_set_client`](/system-variables.md#character_set_client) 和 [`character_set_connection`](/system-variables.md#character_set_connection) 不同时设置为 `gbk` 的情况下，TiDB 处理非法字符的方式与 MySQL 一致。
* 在 `character_set_client` 和 `character_set_connection` 同时为 `gbk` 的情况下， TiDB 处理非法字符的方式与 MySQL 有所区别。

    - MySQL 处理非法 GBK 字符集时，对读和写操作的处理方式不同。
    - TiDB 处理非法 GBK 字符集时，对读和写操作的处理方式相同。TiDB 在严格模式下读写非法 GBK 字符都会报错，在非严格模式下，读写非法 GBK 字符都会用 `?` 替换。

>  

Reviewer 4: ```suggestion
```
